torch_xla/csrc/aten_cpu_fallback.cpp

```diff
@@ -1,34 +1,362 @@
+#include <vector>
 
+#include "torch_xla/csrc/aten_xla_bridge.h"
+#include "torch_xla/csrc/dl_convertor.h"
 #include "torch_xla/csrc/function_call_tracker.h"
 #include "torch_xla/csrc/runtime/debug_macros.h"
 #include "torch_xla/csrc/runtime/metrics.h"
 #include "torch_xla/csrc/runtime/tf_logging.h"
+#include "torch_xla/csrc/xla_graph_executor.h"
+
+// Forward declaration of PyTorch CUDA functions.
+// Source: c10/cuda/CUDAFunctions.h
+//
+// These are needed in order to synchronize the CUDA device after running
+// the operation in PyTorch eager mode.
+//
+// It would be better to include the actual header. However, if we build
+// PyTorch/XLA in an environment where PyTorch wasn't compiled with CUDA
+// (i.e. our CI), the build would fail.
+namespace c10::cuda {
+
+at::DeviceIndex current_device() noexcept;
+void set_device(at::DeviceIndex);
+void device_synchronize();
+
+}  // namespace c10::cuda
 
 namespace torch_xla {
 
 // TODO(jwtan): Replace this with torch::lazy::Counter. We need
-// _cpu_fallback_counters to remain as torch_xla::runtime::metrics::Counter to
+// _fallback_counters to remain as torch_xla::runtime::metrics::Counter to
 // support torch_xla::runtime::metrics::CreatePerformanceReport(). For more
 // information, see NOTE: [TORCH_LAZY_COUNTER v.s. XLA_COUNTER].
 static std::unordered_map<std::string, ::torch_xla::runtime::metrics::Counter*>
-    _cpu_fallback_counters;
+    _fallback_counters;
 
 // Get all the executed fallback operations.
 // In other words, get all of them whose counters are not zero.
 std::vector<std::string> GetFallbackOperations() {
   std::vector<std::string> fallback;
-  for (auto const& pair : _cpu_fallback_counters) {
+  for (auto const& pair : _fallback_counters) {
     if (pair.second->Value() != 0) {
       fallback.push_back(pair.first);
     }
   }
   return fallback;
 }
 
-void xla_cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+// Most of the functions for the CUDA fallback are a modified version of
+// PyTorch's at::native::cpu_fallback function.
+//
+// Source: aten/src/ATen/native/CPUFallback.cpp
+//
+// While a better solution would be to adapt PyTorch's function to be device
+// agnostic, the changes are not small enough that would make sense for adding
+// just one more device. Therefore, we copied the needed functions in this file.
+//
+// Before each modified function below, we shall specify what has changed,
+// if there was any.
+
+bool UseCUDAFallback() {
+  return runtime::sys_util::GetEnvBool("XLA_FALLBACK_CUDA", false);
+}
+
+// Change: use of std::any_of instead of iterating with a for-loop.
+static bool validate_tensor_list(const c10::List<at::Tensor>& tensorlist) {
+  return std::any_of(tensorlist.begin(), tensorlist.end(),
+                     [](const at::Tensor& tensor) { return tensor.defined(); });
+}
+
+// Retrieve the inner XLATensorPtr, and check it lives inside CUDA.
+static XLATensorPtr get_xla_tensor(const at::Tensor& tensor) {
```

Comment:
nit: `get_xla_cuda_tensor` could be a better name since you assert the device to be cuda?